File: src/main/java/com/freya02/botcommands/internal/Cooldownable.java
package com.freya02.botcommands.internal;

import com.freya02.botcommands.api.CooldownScope;
import com.freya02.botcommands.api.Logging;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent;
import net.dv8tion.jda.api.interactions.Interaction;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Supplier;

public abstract class Cooldownable {
	private static final Logger LOGGER = Logging.getLogger();
	private final CooldownStrategy cooldownStrategy;

	//The values is the time on which the cooldown expires
	private final Map<Long, Map<Long, Long>> userCooldowns = Collections.synchronizedMap(new HashMap<>());
	private final Map<Long, Long> channelCooldowns = Collections.synchronizedMap(new HashMap<>());
	private final Map<Long, Long> guildCooldowns = Collections.synchronizedMap(new HashMap<>());

	protected Cooldownable(CooldownStrategy cooldownStrategy) {
		this.cooldownStrategy = cooldownStrategy;
	}

	public long getCooldownMillis() {
		return cooldownStrategy.getCooldownMillis();
	}

	public CooldownScope getCooldownScope() {
		return cooldownStrategy.getScope();
	}

	public void applyCooldown(GuildMessageReceivedEvent event) {
		switch (getCooldownScope()) {
			case USER -> getGuildUserCooldownMap(event.getGuild()).put(event.getAuthor().getIdLong(), System.currentTimeMillis() + getCooldownMillis());
			case GUILD -> guildCooldowns.put(event.getGuild().getIdLong(), System.currentTimeMillis() + getCooldownMillis());
			case CHANNEL -> channelCooldowns.put(event.getChannel().getIdLong(), System.currentTimeMillis() + getCooldownMillis());
		}
	}

	@NotNull
	private Map<Long, Long> getGuildUserCooldownMap(Guild guild) {
		return userCooldowns.computeIfAbsent(guild.getIdLong(), x -> Collections.synchronizedMap(new HashMap<>()));
	}

	public void applyCooldown(Interaction event) {
		switch (getCooldownScope()) {
			case USER -> {
				if (event.getGuild() == null) break;
				getGuildUserCooldownMap(event.getGuild()).put(event.getUser().getIdLong(), System.currentTimeMillis() + getCooldownMillis());
			}
			case GUILD -> {
				if (event.getGuild() == null) break;
				guildCooldowns.put(event.getGuild().getIdLong(), System.currentTimeMillis() + getCooldownMillis());
			}
			case CHANNEL -> {
				if (event.getChannel() == null) break;
				channelCooldowns.put(event.getChannel().getIdLong(), System.currentTimeMillis() + getCooldownMillis());
			}
		}
	}

	public long getCooldown(GuildMessageReceivedEvent event) {
		return switch (getCooldownScope()) {
			case USER -> Math.max(0, getGuildUserCooldownMap(event.getGuild()).getOrDefault(event.getAuthor().getIdLong(), 0L) - System.currentTimeMillis());
			case GUILD -> Math.max(0, guildCooldowns.getOrDefault(event.getGuild().getIdLong(), 0L) - System.currentTimeMillis());
			case CHANNEL -> Math.max(0, channelCooldowns.getOrDefault(event.getChannel().getIdLong(), 0L) - System.currentTimeMillis());
		};
	}

	public long getCooldown(Interaction event, Supplier<String> cmdNameSupplier) {
		switch (getCooldownScope()) {
			case USER -> {
				if (event.getGuild() == null) {
					LOGGER.warn("Interaction command '{}' wasn't used in a guild but uses a guild-wide cooldown", cmdNameSupplier.get());

					return 0;
				}

				return Math.max(0, getGuildUserCooldownMap(event.getGuild()).getOrDefault(event.getUser().getIdLong(), 0L) - System.currentTimeMillis());
			}
			case GUILD -> {
				if (event.getGuild() == null) {
					LOGGER.warn("Interaction command '{}' wasn't used in a guild but uses a guild-wide cooldown", cmdNameSupplier.get());

					return 0;
				}

				return Math.max(0, guildCooldowns.getOrDefault(event.getGuild().getIdLong(), 0L) - System.currentTimeMillis());
			}
			case CHANNEL -> {
				if (event.getChannel() == null) {
					LOGGER.warn("Interaction command '{}' wasn't used in a channel, somehow", cmdNameSupplier.get());

					return 0;
				}

				return Math.max(0, channelCooldowns.getOrDefault(event.getChannel().getIdLong(), 0L) - System.currentTimeMillis());
			}
			default -> throw new IllegalStateException("Unexpected value: " + getCooldownScope());
		}
	}
}

File: src/main/java/com/freya02/botcommands/internal/Cooldownable.kt
package com.freya02.botcommands.internal

import com.freya02.botcommands.api.CooldownScope
import gnu.trove.TCollections
import gnu.trove.impl.Constants
import gnu.trove.map.hash.TObjectLongHashMap
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import net.dv8tion.jda.api.events.interaction.command.GenericCommandInteractionEvent
import net.dv8tion.jda.api.events.message.MessageReceivedEvent

sealed class Cooldownable(private val context: BContextImpl, cooldownStrategy: CooldownStrategy) {
    @JvmRecord
    private data class CooldownKey(private val placeId: Long?, private val userId: Long)

    //Trove sets are not *always* the fastest, by a small margin, but are by far the most memory efficient
    private val cooldowns =
        TCollections.synchronizedMap(TObjectLongHashMap<CooldownKey>(Constants.DEFAULT_CAPACITY, Constants.DEFAULT_LOAD_FACTOR, 0))

    private val cooldownMillis: Long = cooldownStrategy.cooldownMillis
    val cooldownScope: CooldownScope = cooldownStrategy.scope

    fun applyCooldown(event: MessageReceivedEvent) = applyCooldown(event.toCooldownKey())

    fun applyCooldown(event: GenericCommandInteractionEvent) = applyCooldown(event.toCooldownKey())

    private fun applyCooldown(cooldownKey: CooldownKey) {
        cooldowns.put(cooldownKey, System.currentTimeMillis() + cooldownMillis)
        context.config.coroutineScopesConfig.cooldownScope.launch {
            delay(cooldownMillis)
            cooldowns.remove(cooldownKey)
        }
    }

    fun getCooldown(event: MessageReceivedEvent) = cooldowns[event.toCooldownKey()] - System.currentTimeMillis()

    fun getCooldown(event: GenericCommandInteractionEvent) = cooldowns[event.toCooldownKey()] - System.currentTimeMillis()

    private fun MessageReceivedEvent.toCooldownKey(): CooldownKey {
        if (!isFromGuild) throwInternal("Invalid cooldown scope for text commands")
        return when (cooldownScope) {
            CooldownScope.USER -> CooldownKey(guild.idLong, author.idLong)
            CooldownScope.GUILD -> CooldownKey(guild.idLong, author.idLong)
            CooldownScope.CHANNEL -> CooldownKey(channel.idLong, author.idLong)
        }
    }

    private fun GenericCommandInteractionEvent.toCooldownKey() = when (cooldownScope) {
        CooldownScope.USER -> {
            CooldownKey(guild?.idLong, user.idLong)
        }

        CooldownScope.GUILD -> {
            val guild = guild ?: throwInternal("Invalid cooldown scope for text commands")
            CooldownKey(guild.idLong, user.idLong)
        }

        CooldownScope.CHANNEL -> {
            if (!isFromGuild) throwInternal("Invalid cooldown scope for text commands")
            CooldownKey(guildChannel.idLong, user.idLong)
        }
    }
}